desktop/test/e2e/specs/login.js

```diff
@@ -49,20 +51,36 @@ describe( 'User Can log in', () => {
 			},
 			env: {
 				...process.env,
-				WP_DESKTOP_BASE_URL: process.env.WP_DESKTOP_BASE_URL?.replace( /\/$/, '' ) ?? undefined,
+				WP_DESKTOP_BASE_URL: BASE_URL,
 				WP_DEBUG_LOG, // This will override logging path from the Electron main process.
 				// Ensure other CI-specific overrides (such as disabling the auto-updater)
 				DEBUG: true,
 				CI: true,
 			},
 		} );
+
+		// Find main window. Playwright has problems identifying the main window when using `firstWindow`, so we
+		// iterate over all windows and find it by URL.
+		for ( const window of await electronApp.windows() ) {
```

Comment:
Interesting - does this change seem to make performance more consistent? From my understanding there is only a single window object, and the element in which WordPress/Calypso is loaded is essentially a "frame" in that object. You can verify this by examining the following:

```
for ( const [ _, frame ] of mainWindow.frames().entries() ) {
  console.log( 'Frame title: ', frame.title();
  console.log( 'Frame URL:', frame.url();
}
```

However it's also quite possible that perhaps Playwright sees as different "windows" as well, so doesn't hurt to try any and all ideas.

On a related note, it just occurred to me to check whether we're structuring the login test in the same way as in Calypso? For example, in my testing, _timing_ seemed to be the main issue - for some reason, Playwright was able to successfully wait for specific selectors during some trials, and failed to latch onto the same selectors during others. Even though these selectors are in fact always present when inspected via developer tools! Really odd. 🤔 The TL;DR of my question here is perhaps the Calypso login test waits/selects elements in a very specific way that avoids this problem? Maybe there's some other elements we're missing. Not sure but might be worth double-checking. (Although if Calypso isn't using Playwright at all for the same tests this point is moot, I suppose.)

Reply:
Looking at Playwright code, main window is either literally the first window, or the first object that throws the event `window`. In that context, a "window" is a list of objects that throw the event `page`.

I don't know if those events are Playwright specific or something more generic. But I guess that given it's based on events there is always the chance of race conditions, it sounds brittle to me.

Reply:
Calypso uses https://github.com/Automattic/wp-calypso/blob/74d1739c394880525967228e82bc835f5ba7cdbd/packages/calypso-e2e/src/lib/pages/login-page.ts#L102 to exercise the Login window. The version I implemented in this test is a super-simplified version, mostly to prove it works.

I'd love to use the same library used by Calypso, but for that to work first, we need to migrate Desktop to yarn v 3.1, and then integrate it with the main dependency tree so it has visibility on `@automattic/calypso-e2e`. First step of that journey is in https://github.com/Automattic/wp-calypso/pull/57487

Reply:
Re: stability. After this change I had 5 green runs in a row in TeamCity, so it seems to be helping!

Reply:
Re: Re: stability. I'm getting lots of green builds in CircleCI as well. I'm going to clean up the code a bit so it is ready to be merged.